File: sdk/python/aegis_sdk/worker.py
"""Python worker SDK for Aegis.

Connects to an Aegis server's WebSocket dispatch endpoint, registers itself
with the job types it can handle, and executes jobs pushed to it — handling
concurrency limiting, periodic heartbeats for long-running jobs, structured
completion/failure reporting, and automatic reconnect with backoff.
"""

from __future__ import annotations

import asyncio
import inspect
import json
import logging
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Optional, Union

import websockets

logger = logging.getLogger("aegis_sdk")

Handler = Callable[["Job"], Union[Optional[dict], Awaitable[Optional[dict]]]]


@dataclass
class Job:
    """A unit of work dispatched to this worker."""

    id: str
    type: str
    payload: Any
    attempts: int = 0
    max_attempts: int = 3
    priority: int = 0
    queue: str = "default"
    workflow_id: Optional[str] = None
    workflow_step: Optional[str] = None
    raw: dict = field(default_factory=dict)

    @classmethod
    def from_wire(cls, data: dict) -> "Job":
        return cls(
            id=data["id"],
            type=data["type"],
            payload=data.get("payload"),
            attempts=data.get("attempts", 0),
            max_attempts=data.get("max_attempts", 3),
            priority=data.get("priority", 0),
            queue=data.get("queue", "default"),
            workflow_id=data.get("workflow_id") or None,
            workflow_step=data.get("workflow_step") or None,
            raw=data,
        )


class Worker:
    """Connects to an Aegis server and executes jobs using registered handlers."""

    def __init__(
        self,
        server_addr: str,
        name: str,
        queues: Optional[list[str]] = None,
        concurrency: int = 4,
        heartbeat_every: float = 15.0,
    ):
        self.server_addr = server_addr
        self.name = name
        self.queues = queues or ["default"]
        self.concurrency = concurrency
        self.heartbeat_every = heartbeat_every
        self._handlers: dict[str, Handler] = {}

    def handle(self, job_type: str):
        """Decorator registering a handler for `job_type`.

        The handler may be sync or async, and may return a JSON-serializable
        dict (or None) as the job result. Raising fails the job; Aegis
        retries with backoff up to max_attempts before dead-lettering it.
        """

        def decorator(fn: Handler) -> Handler:
            self._handlers[job_type] = fn
            return fn

        return decorator

    async def run(self) -> None:
        """Connect and process jobs forever, reconnecting with backoff on
        any connection failure. Runs until cancelled."""
        backoff = 1.0
        while True:
            try:
                await self._run_once()
                backoff = 1.0
            except (websockets.exceptions.ConnectionClosed, OSError) as exc:
                logger.warning("aegis worker disconnected (%s), retrying in %.1fs", exc, backoff)
                await asyncio.sleep(backoff)
                backoff = min(backoff * 2, 30.0)

    async def _run_once(self) -> None:
        async with websockets.connect(self.server_addr, max_size=8 * 1024 * 1024) as ws:
            send_lock = asyncio.Lock()

            async def send(msg: dict) -> None:
                async with send_lock:
                    await ws.send(json.dumps(msg))

            await send(
                {
                    "type": "register",
                    "name": self.name,
                    "queues": self.queues,
                    "job_types": list(self._handlers.keys()),
                    "concurrency": self.concurrency,
                }
            )

            active: dict[str, bool] = {}
            heartbeat_task = asyncio.create_task(self._heartbeat_loop(send, active))
            try:
                await send({"type": "request", "count": self.concurrency})
                async for raw in ws:
                    msg = json.loads(raw)
                    msg_type = msg.get("type")

                    if msg_type == "job":
                        job = Job.from_wire(msg["job"])
                        active[job.id] = True
                        asyncio.create_task(self._handle_job(send, job, active))
                    elif msg_type == "registered":
                        logger.info("aegis worker registered: worker_id=%s name=%s", msg.get("worker_id"), self.name)
                    elif msg_type == "error":
                        logger.error("aegis server error: %s", msg.get("message"))
            finally:
                heartbeat_task.cancel()

    async def _heartbeat_loop(self, send, active: dict[str, bool]) -> None:
        try:
            while True:
                await asyncio.sleep(self.heartbeat_every)
                for job_id in list(active.keys()):
                    await send({"type": "heartbeat", "job_id": job_id})
        except asyncio.CancelledError:
            pass

    async def _handle_job(self, send, job: Job, active: dict[str, bool]) -> None:
        handler = self._handlers.get(job.type)
        try:
            if handler is None:
                raise RuntimeError(f"no handler registered for job type {job.type!r}")
            result = handler(job)
            if inspect.isawaitable(result):
                result = await result
            await send({"type": "complete", "job_id": job.id, "result": result or {}})
        except Exception as exc:  # noqa: BLE001 - reporting the failure to the server is the point
            logger.exception("job %s (%s) failed", job.id, job.type)
            await send({"type": "fail", "job_id": job.id, "error": str(exc)})
        finally:
            active.pop(job.id, None)
            await send({"type": "request", "count": 1})

File: sdk/go/worker.go
// Package aegis is the Go worker SDK: register handlers for job types,
// connect to an Aegis server over its WebSocket dispatch protocol, and let
// the SDK handle registration, concurrency limiting, heartbeats (so
// long-running jobs don't get reclaimed as dead), graceful failure
// reporting, and automatic reconnect with backoff.
//
// Minimal usage:
//
//	w := aegis.NewWorker("ws://localhost:8080/ws/worker", "image-workers")
//	w.Handle("resize_image", func(ctx context.Context, job *aegis.Job) (json.RawMessage, error) {
//	    // ... do work using job.Payload ...
//	    return json.RawMessage(`{"ok":true}`), nil
//	})
//	w.Run(context.Background())
package aegis

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"github.com/gorilla/websocket"

	"github.com/Heet852003/aegis/internal/models"
)

// Job is the unit of work handed to a Handler. It is the same shape the
// server persists, so handlers can inspect attempt counts, priority, etc.
type Job = models.Job

// Handler executes one job and returns either a JSON-serializable result or
// an error. Returning an error fails the job, which the server will retry
// with backoff (up to the job's MaxAttempts) before dead-lettering it —
// handlers don't need their own retry logic.
type Handler func(ctx context.Context, job *Job) (json.RawMessage, error)

type wireMsg struct {
	Type        string          `json:"type"`
	Name        string          `json:"name,omitempty"`
	Queues      []string        `json:"queues,omitempty"`
	JobTypes    []string        `json:"job_types,omitempty"`
	Concurrency int             `json:"concurrency,omitempty"`
	Count       int             `json:"count,omitempty"`
	WorkerID    string          `json:"worker_id,omitempty"`
	Job         *Job            `json:"job,omitempty"`
	JobID       string          `json:"job_id,omitempty"`
	Result      json.RawMessage `json:"result,omitempty"`
	Error       string          `json:"error,omitempty"`
	Message     string          `json:"message,omitempty"`
}

// Worker connects to an Aegis server and executes jobs using registered
// Handlers.
type Worker struct {
	ServerAddr     string // e.g. "ws://localhost:8080/ws/worker"
	Name           string
	Queues         []string      // defaults to ["default"]
	Concurrency    int           // max jobs in flight at once; defaults to 4
	HeartbeatEvery time.Duration // defaults to 15s

	handlers map[string]Handler
	mu       sync.Mutex
}

func NewWorker(serverAddr, name string) *Worker {
	return &Worker{
		ServerAddr:  serverAddr,
		Name:        name,
		Queues:      []string{"default"},
		Concurrency: 4,
		handlers:    make(map[string]Handler),
	}
}

// Handle registers fn as the handler for jobType. Call before Run.
func (w *Worker) Handle(jobType string, fn Handler) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.handlers[jobType] = fn
}

func (w *Worker) jobTypes() []string {
	w.mu.Lock()
	defer w.mu.Unlock()
	types := make([]string, 0, len(w.handlers))
	for t := range w.handlers {
		types = append(types, t)
	}
	return types
}

func (w *Worker) handlerFor(jobType string) (Handler, bool) {
	w.mu.Lock()
	defer w.mu.Unlock()
	h, ok := w.handlers[jobType]
	return h, ok
}

// Run connects and processes jobs until ctx is cancelled, automatically
// reconnecting with backoff if the connection drops.
func (w *Worker) Run(ctx context.Context) error {
	if w.Concurrency <= 0 {
		w.Concurrency = 4
	}
	if w.HeartbeatEvery <= 0 {
		w.HeartbeatEvery = 15 * time.Second
	}
	backoff := time.Second
	for {
		if ctx.Err() != nil {
			return ctx.Err()
		}
		err := w.runOnce(ctx)
		if ctx.Err() != nil {
			return ctx.Err()
		}
		slog.Warn("aegis worker disconnected, retrying", "error", err, "backoff", backoff)
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(backoff):
		}
		if backoff < 30*time.Second {
			backoff *= 2
		}
	}
}

func (w *Worker) runOnce(ctx context.Context) error {
	ws, _, err := websocket.DefaultDialer.DialContext(ctx, w.ServerAddr, nil)
	if err != nil {
		return fmt.Errorf("dial: %w", err)
	}
	defer ws.Close()

	var writeMu sync.Mutex
	send := func(m wireMsg) error {
		writeMu.Lock()
		defer writeMu.Unlock()
		return ws.WriteJSON(m)
	}

	if err := send(wireMsg{
		Type:        "register",
		Name:        w.Name,
		Queues:      w.Queues,
		JobTypes:    w.jobTypes(),
		Concurrency: w.Concurrency,
	}); err != nil {
		return fmt.Errorf("register: %w", err)
	}

	sessionCtx, cancel := context.WithCancel(ctx)
	defer cancel()

	active := &sync.Map{} // jobID -> struct{} while running, used for heartbeats
	var inflight sync.WaitGroup

	go func() {
		ticker := time.NewTicker(w.HeartbeatEvery)
		defer ticker.Stop()
		for {
			select {
			case <-sessionCtx.Done():
				return
			case <-ticker.C:
				active.Range(func(k, _ any) bool {
					send(wireMsg{Type: "heartbeat", JobID: k.(string)})
					return true
				})
			}
		}
	}()

	// Ask for our first batch of work immediately.
	if err := send(wireMsg{Type: "request", Count: w.Concurrency}); err != nil {
		return fmt.Errorf("initial request: %w", err)
	}

	for {
		var msg wireMsg
		if err := ws.ReadJSON(&msg); err != nil {
			cancel()
			inflight.Wait()
			return fmt.Errorf("read: %w", err)
		}

		switch msg.Type {
		case "registered":
			slog.Info("aegis worker registered", "worker_id", msg.WorkerID, "name", w.Name)

		case "job":
			job := msg.Job
			handler, ok := w.handlerFor(job.Type)
			if !ok {
				send(wireMsg{Type: "fail", JobID: job.ID, Error: fmt.Sprintf("no handler registered for job type %q", job.Type)})
				send(wireMsg{Type: "request", Count: 1})
				continue
			}
			active.Store(job.ID, struct{}{})
			inflight.Add(1)
			go func() {
				defer inflight.Done()
				defer active.Delete(job.ID)
				result, err := handler(sessionCtx, job)
				if err != nil {
					send(wireMsg{Type: "fail", JobID: job.ID, Error: err.Error()})
				} else {
					send(wireMsg{Type: "complete", JobID: job.ID, Result: result})
				}
				send(wireMsg{Type: "request", Count: 1})
			}()

		case "error":
			slog.Error("aegis server error", "message", msg.Message)
		}
	}
}
